gobblin-core/src/main/java/gobblin/source/extractor/partition/Partitioner.java

```diff
@@ -33,26 +39,49 @@
 import gobblin.source.extractor.watermark.WatermarkPredicate;
 import gobblin.source.extractor.watermark.WatermarkType;
 
-
 /**
  * An implementation of default partitioner for all types of sources
  */
 public class Partitioner {
   private static final String WATERMARKTIMEFORMAT = "yyyyMMddHHmmss";
   private static final Logger LOG = LoggerFactory.getLogger(Partitioner.class);
+  public static final String HAS_USER_SPECIFIED_HIGH_WATERMARK = "has.user.specified.high.watermark";
+
+  public static final Comparator<Partition> ascendingComparator = new Comparator<Partition>() {
+    @Override
+    public int compare(Partition p1, Partition p2) {
+      if (p1 == null && p2 == null) {
+        return 0;
+      }
+      if (p1 == null) {
+        return -1;
+      }
+      if (p2 == null) {
+        return 1;
+      }
+      return p1.getLowWatermark() > p2.getLowWatermark() ? 1 : -1;
```

Comment:
Should add p1.getLowWatermark() == p2.getLowWatermark() case for completeness. I think that can just be merged with the first if check.